### DISCUSSION TITLE
[v4] Performance issues

### DISCUSSION THREAD

**Discussion by atanaskanchev** [nativewind/nativewind#642] [General]
**Describe the bug**
Running a [performance test](https://github.com/efstathiosntonas/react-native-style-libraries-benchmark/pull/9) of the latest versions of some popular UI frameworks it seems like NativeWind V4 has some performance issues. The benchmark renders 1000 styled View components and NW4 is 400% slower than Stylesheet

**To Reproduce**
Checkout the benchmark and run it

**Screenshots**
Bare React Native
<img src='https://github.com/marklawlor/nativewind/assets/4958587/db31d510-0da7-4066-88f7-b6bef0fc715c' width='350'>

NW4
<img src='https://github.com/marklawlor/nativewind/assets/4958587/b0cd519b-b954-4d81-8e11-27526eb579b1' width='350'>

**Comment by ioss** [comment]
Just as a note: I can confirm that on a plain expo-router project in a production build on an iPhone. (without any of the other styling libraries)

**Comment by marklawlor** [comment]
For the initial release of v4 this is to be expected. I'm the only maintainer and I do not have enough time to work on both new features and performance. Post-release you expect performance improvements aiming to return to v2 performance speeds.

I also don't like that benchmark. Its a very narrow-focused micro-benchmark. I'd like to see a benchmark with things like interactivity / screen transitions / etc.

I'm going to convert this to a discussion as there's no end to this issue. It will always be an ongoing topic

**Comment by marklawlor** [comment]
The biggest bottleneck in performance is due to a bug in `react-native-reanimated`

Reported: https://github.com/software-mansion/react-native-reanimated/issues/5296
In-progress PR: https://github.com/software-mansion/react-native-reanimated/pull/5268

Due to this bug, inline animated styles needs to be wrapped in `useAnimatedStyle()` for them to function. Unfortunately, this means using a hook which must obey the rules of hooks forcing us to make every component "animated".

https://github.com/marklawlor/nativewind/blob/ebd38d250ef4f4bb0748d1c75d3509978bb903b8/packages/react-native-css-interop/src/runtime/css-interop.native.ts#L54-L87

Using the supplied benchmark, removing these lines reduces the render time to just slower than GlueStack.

I'm welcome to ideas on how to fix this, but I'd like something that works for all edge cases (e.g dynamically adding & removing animations, animations working through fast-refresh).

Also, the solution isn't as easy as simply "can't you use the `isAnimated` flag to conditionally render?". Yes and no. I tried. The problem is during development, components can easily switch between being animated and non-animated. E.g just writing `<Text className="text-red-500" />`, and then you decide to change it to `<Text className="text-red-500 transition" />`. It will fast-refresh and you'll get an error about hooks. You cannot get rid of this error until you reload the app. I've found that people trigger this error multiple times and are constantly fighting the error messages :cry: 

So, if fast-refresh is the issue - just do it in production? I don't know if crashing in production is a good solution.

That's why the current solution is "everything is animated" - at least until `react-native-reanimated` fixes the issue.

**Reply by nandorojo** [reply]
@marklawlor not sure if this is still an issue or not...but if it is, I think it's better to break fast refresh than have worse runtime perf. Tamagui has the same issue, and it also breaks when you add an animation in dev mode. 

One doable solution: when in dev mode, listen for changes in `isAnimated` and have a clearer warning when it happens.

It's an unfortunate trade off but reanimated perf has gotten super heavy for render times.

I can peak into the code to see if it's still a thing.

**Reply by nandorojo** [reply]
I'm happy to discuss this more and see if I can help if needed

**Reply by claudesortwell** [reply]
@nandorojo Isnt this already done? https://github.com/nativewind/nativewind/blob/4802c5c9adb0923846e4296e4816195d1ff8e77f/packages/react-native-css-interop/src/runtime/native/render-component.tsx#L69

Seems like useAnimated is only used if there is an animation style on the on component

**Comment by hari-mohan-choudhary** [comment]
Now, this PR has already been merged: https://github.com/software-mansion/react-native-reanimated/pull/5268. @marklawlor, has it been improved as per your suggestion?

**Comment by markwitt1** [comment]
Any update on this? Any way to help out?

**Comment by roo12312** [comment]
even im facing performance issues on complex ui.. v2 was better i guess

**Comment by danstepanov** [comment]
you should see some performance improvements with Nativewind v5 (now available) and we will be focused on improving performance now that nativewind and react-native-css have been separated.
